### PR TITLE
Issue 1992: Compute unread bytes based on the last checkpoint.

### LINKED ISSUE #1992: Bug fix and test case for ReaderGroup.unreadBytes
**Problem description**
I couldn't find any unit test or integration test exercising `ReaderGroup.unreadBytes`. While writing one, I got an NPE, which I fixed by adding a call to `fetchUpdates` in the implementation of the `unreadBytes` calls. The next problem I faced is that the call to `getSuccessors` given a stream cut is not implemented in `LocalController`.

**Problem location**
Client, tests.

**Suggestions for an improvement**
Add test cases.

### PR DESCRIPTION
**Pull request by shrids**
Signed-off-by: shrids <sandeep.shridhar@emc.com>
**Change log description**
Compute `unreadBytes()` based on segments and offsets of the last successful checkpoint. 

**Purpose of the change**
Fixes #1992

**What the code does**
- Ensure `unreadBytes()` uses the segments and offsets of the last checkpoint for computing unreadBytes.
- If there are no checkpoints then use the offsets of segments at the time they were assigned to the readers. (older method of computation.)

**How to verify it**
Tests should continue to pass.

### CODE REVIEW

File: test/integration/src/test/java/io/pravega/test/integration/UnreadBytesTest.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    @After
+    public void tearDown() throws Exception {
+        executor.shutdown();
+        controllerWrapper.close();
+        server.close();
+        serviceBuilder.close();
+        zkTestServer.close();
+    }
+
+    @Test//(timeout = 40000)
```
Comment (tkaitchuck) [inline test/integration/src/test/java/io/pravega/test/integration/UnreadBytesTest.java]: uncomment

File: client/src/main/java/io/pravega/client/stream/impl/CheckpointState.java [hunk 2]
```diff
@@ -111,7 +120,7 @@ void clearCheckpointsThrough(String checkpointId) {
             for (Iterator<String> iterator = checkpoints.iterator(); iterator.hasNext();) {
                 String cp = iterator.next();
                 uncheckpointedHosts.remove(cp);
-                checkpointPositions.remove(cp);
+                lastCheckpointPosition = checkpointPositions.remove(cp);
```
[the remark this answers is not in the document]
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/impl/CheckpointState.java]: done.

File: test/integration/src/test/java/io/pravega/test/integration/UnreadBytesTest.java (hunk 1 above)
Comment (shrids) [reply test/integration/src/test/java/io/pravega/test/integration/UnreadBytesTest.java]: done.

**Comment by shrids** [comment]
Successfully ran system tests (`:test:system`) with the changes (Build: `test-pravega-continuous #873`)

File: client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java [hunk 3]
```diff
@@ -189,17 +190,29 @@ public ReaderGroupMetrics getMetrics() {
...
+
+        Optional<Map<Stream, Map<Segment, Long>>> checkPointedPositions =
+                synchronizer.getState().getPositionsForLastCompletedCheckpoint();
         SegmentMetadataClientFactory metaFactory = new SegmentMetadataClientFactoryImpl(controller, connectionFactory);
-        
+        if (checkPointedPositions.isPresent()) {
+            log.info("Computing unread bytes based on the last checkPoint position");
+            return getUnreadBytes(checkPointedPositions.get(), metaFactory);
+        } else {
+            log.warn("No checkpoints found, using the last known offset to compute unread bytes");
```
[the remark this answers is not in the document]
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java]: Corrected it.

File: client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java [hunk 4]
```diff
@@ -189,17 +190,29 @@ public ReaderGroupMetrics getMetrics() {
...
         StateSynchronizer<ReaderGroupState> synchronizer = createSynchronizer();
-        Map<Stream, Map<Segment, Long>> positions = synchronizer.getState().getPositions();
+        synchronizer.fetchUpdates();
+
+        Optional<Map<Stream, Map<Segment, Long>>> checkPointedPositions =
+                synchronizer.getState().getPositionsForLastCompletedCheckpoint();
         SegmentMetadataClientFactory metaFactory = new SegmentMetadataClientFactoryImpl(controller, connectionFactory);
-        
+        if (checkPointedPositions.isPresent()) {
+            log.info("Computing unread bytes based on the last checkPoint position");
```
[the remark this answers is not in the document]
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java]: Updated it.

File: client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java [hunk 5]
```diff
@@ -189,17 +190,29 @@ public ReaderGroupMetrics getMetrics() {
...
+            return getUnreadBytes(checkPointedPositions.get(), metaFactory);
+        } else {
+            log.warn("No checkpoints found, using the last known offset to compute unread bytes");
+            return getUnreadBytes(synchronizer.getState().getPositions(), metaFactory);
+        }
+    }
+
+    private long getUnreadBytes(Map<Stream, Map<Segment, Long>> positions, SegmentMetadataClientFactory metaFactory) {
```
[the remark this answers is not in the document]
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java]: Added additional log message.

**Pull request merged**